pkg/ip/ip.go

```diff
@@ -85,31 +85,97 @@ func IsIP(version types.IPVersion, s string) error {
 //
 // If sorted is true, the result set of IP addresses will be sorted.
 func IPsDiffSet(ips1, ips2 []net.IP, sorted bool) []net.IP {
-	var ips []net.IP
-	marks := make(map[string]bool)
-	for _, ip := range ips1 {
+	return getIPDiffSet(ips1, ips2, sorted, -1)
+}
+
+func IsDiffIPSet(ips1, ips2 []net.IP) bool {
+	diff := getIPDiffSet(ips1, ips2, false, 1)
+	return len(diff) > 0
+}
+
+func getIPDiffSet(ips1, ips2 []net.IP, sorted bool, want int) []net.IP {
+	ips2Map := make(map[[16]byte]struct{}, len(ips2))
+	for _, ip := range ips2 {
 		if ip != nil {
-			marks[ip.String()] = true
+			ips2Map[[16]byte(ip.To16())] = struct{}{}
 		}
 	}
 
-	for _, ip := range ips2 {
+	var result []net.IP
+	for _, ip := range ips1 {
 		if ip != nil {
-			delete(marks, ip.String())
+			if _, ok := ips2Map[[16]byte(ip.To16())]; !ok {
+				result = append(result, ip)
+				if want > 0 && len(result) >= want {
+					break
+				}
+			}
 		}
 	}
 
-	for k := range marks {
-		ips = append(ips, net.ParseIP(k))
+	if sorted && len(result) > 1 {
+		sort.Slice(result, func(i, j int) bool {
+			return bytes.Compare(result[i].To16(), result[j].To16()) < 0
+		})
 	}
 
-	if sorted {
-		sort.Slice(ips, func(i, j int) bool {
-			return bytes.Compare(ips[i].To16(), ips[j].To16()) < 0
-		})
+	return result
+}
+
+// FindAvailableIPs find available ip list in range
+func FindAvailableIPs(ipRanges []string, ipList []net.IP, count int) []net.IP {
```

Comment:
paramenter:
ipRanges 
ipList -> excludeIPList
count -> expectCount

comment: find expectCount IP address from ipRanges, but exclude from excludeIPList

Reply:
optimize the parameter name